File: app/src/main/java/com/lapislucera/wheresthemoon/WTMAppWidget.java
package com.lapislucera.wheresthemoon;

import android.annotation.SuppressLint;
import android.app.AlarmManager;
import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.Context;
import android.content.Intent;
import android.database.Cursor;
import android.graphics.Typeface;
import android.os.SystemClock;
import android.util.TypedValue;
import android.widget.RemoteViews;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Color;
import java.text.SimpleDateFormat;
import java.util.Calendar;

/**
 * Implementation of App Widget functionality.
 *
 */
public class WTMAppWidget extends AppWidgetProvider {

    static {
        System.loadLibrary("witm");
    }

    /**
     * Converts decimal degrees to degrees, minutes, and seconds
     * @param decimal Longitude in decimal format
     * @return String representations of degrees, minutes, seconds
     */
    private static String decimal2dms( Double decimal ) {

        Integer deg, min, sec;

        deg = (int)Math.floor(decimal);
        min = (int)Math.floor( ( decimal - deg ) * 60 );
        sec = (int)Math.round( ( ( ( decimal - deg ) * 60 ) - min ) * 60 );
        return deg.toString() + "°" + min.toString() + "'" + sec.toString() + "\"";
    }

    public static void updateMoonInfo(final Context context, AppWidgetManager appWidgetManager, int appWidgetId) {

        // variable declarations and initializations
        String      phase_file;
        Bitmap      bmp;
        RemoteViews views         = new RemoteViews(context.getPackageName(), R.layout.wtmapp_widget);
        String[]    signs         = {"q","w","e","r","t","z","u","i","o","p","ü","+"},
                    planets       = { "a","s","d","f","h","j","k","ö","ä","#"},
                    aspects       = {"<","x","c","Q","m"};
        Double      moon_lon      = witm(),
                    sun_lon       = wits();
        Integer     moon_sign     = (int)Math.floor(moon_lon/30.0),
                    label_font_sz = 14,
                    text_font_sz  = 12, phase_lon, moon_icon_id;
        Typeface    kairon        = Typeface.createFromAsset(context.getAssets(),"fonts/KaironSemiserif.ttf"),
                    alegreya      = Typeface.createFromAsset(context.getAssets(),"fonts/AlegreyaSansSC-Regular.ttf"),
                    alegreya_bold = Typeface.createFromAsset(context.getAssets(),"fonts/AlegreyaSansSC-Bold.ttf");
        @SuppressLint("SimpleDateFormat") SimpleDateFormat df = new SimpleDateFormat("MMM d, h:mm a");

        // calculate the phase of the moon and set the moon icon
        if ( sun_lon > moon_lon ) {
            phase_lon = (int)Math.round( 360.0 - sun_lon + moon_lon );
        } else {
            phase_lon = (int)Math.round(moon_lon - sun_lon);
        }
        phase_lon = 18 * (int)Math.floor(phase_lon / 18) + 9;
        phase_file = "moon_" + String.format("%03d", phase_lon);
        moon_icon_id = context.getResources().getIdentifier( phase_file, "drawable", context.getPackageName());
        views.setImageViewResource(R.id.moon_icon,moon_icon_id);

        // set the moon sign and degree
        bmp = getStringBitmap(context, signs[moon_sign] + " ", kairon, text_font_sz );
        views.setImageViewBitmap(R.id.moon_sign, bmp);
        bmp = getStringBitmap(context, decimal2dms(moon_lon % 30), alegreya, text_font_sz );
        views.setImageViewBitmap(R.id.moon_degree, bmp);

        // get the VoC info from the database
        VOCDataBase voc_db = new VOCDataBase(context);
        Cursor   voc_info = voc_db.getCurrentVOC();
        String   sign     = signs[voc_info.getInt(0)],
                 aspect   = aspects[voc_info.getInt(2)],
                 planet   = planets[voc_info.getInt(3)];
        Calendar now      = Calendar.getInstance(),
                 ingress  = Calendar.getInstance(),
                 asptime  = Calendar.getInstance();

        // set the ingress and last aspect times
        ingress.setTimeInMillis(voc_info.getLong(1) * 1000);
        asptime.setTimeInMillis(voc_info.getLong(4) * 1000);

        // round times down to the nearest minute
        ingress.set(Calendar.MILLISECOND, 0);
        ingress.set(Calendar.SECOND, 0);
        asptime.set(Calendar.MILLISECOND, 0);
        asptime.set(Calendar.SECOND, 0);

        // set the labels for the last aspect and ingress sections
        bmp = getStringBitmap(context, "Last Aspect", alegreya_bold, label_font_sz );
        views.setImageViewBitmap(R.id.last_aspect_label, bmp);
        bmp = getStringBitmap(context, "Ingress", alegreya_bold, label_font_sz );
        views.setImageViewBitmap(R.id.ingress_label, bmp);

        // Display ingress sign and datetime
        bmp = getStringBitmap(context, sign + " ", kairon, text_font_sz );
        views.setImageViewBitmap(R.id.ingress_symbol, bmp);
        bmp = getStringBitmap(context, df.format(ingress.getTime()), alegreya, text_font_sz );
        views.setImageViewBitmap(R.id.ingress_time, bmp);

        // Display last aspect sign and datetime
        bmp = getStringBitmap(context, aspect + " " + planet + " ", kairon, text_font_sz );
        views.setImageViewBitmap(R.id.last_aspect_symbols, bmp);
        bmp = getStringBitmap(context, df.format(asptime.getTime()), alegreya, text_font_sz );
        views.setImageViewBitmap(R.id.last_aspect_time, bmp);

        // Is the moon void now?
        if ( now.getTimeInMillis() >= asptime.getTimeInMillis() && now.getTimeInMillis() < ingress.getTimeInMillis() ) {
            // the moon is void
            bmp = getStringBitmap(context, "The moon is void.", alegreya, text_font_sz );
            views.setImageViewBitmap(R.id.void_or_not, bmp);
        } else {
            // the moon is not void
            bmp = getStringBitmap(context, "The moon is not void.", alegreya, text_font_sz );
            views.setImageViewBitmap(R.id.void_or_not, bmp);
        }

        // Instruct the widget manager to update the widget
        appWidgetManager.updateAppWidget(appWidgetId, views);
    }

    private static Bitmap getStringBitmap(Context c, String str, Typeface tf, Integer font_size_in_dp ) {

        // convert font size in dp to pixels
        Integer font_size = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, font_size_in_dp, c.getResources().getDisplayMetrics());

        // create a Paint object and set text properties
        Paint p = new Paint();
        p.setTypeface(tf);
        p.setTextSize(font_size);
        p.setAntiAlias(true);
        p.setSubpixelText(true);
        p.setStyle(Paint.Style.FILL);
        p.setColor(Color.WHITE);
        p.setTextAlign(Paint.Align.CENTER);

        // get the width of the text
        Integer w = (int)p.measureText(str);

        Bitmap bmp = Bitmap.createBitmap(w, font_size, Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(bmp);
        canvas.drawText(str, w / 2, font_size / 2 + (font_size / 4), p);
        return bmp;
    }

    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {

        final AlarmManager m = (AlarmManager)context.getSystemService(Context.ALARM_SERVICE);
        final PendingIntent la_svc, in_svc;
        final Calendar now  = Calendar.getInstance(),
                       la   = Calendar.getInstance(), // last aspect time
                       in   = Calendar.getInstance(); // ingress time
        final Intent   la_i = new Intent(context,WTMService.class).setAction(Intent.ACTION_EDIT),
                       in_i = new Intent(context,WTMService.class).setAction(Intent.ACTION_MAIN);
        final long la_millis, in_millis;

        // get the current VOC info
        VOCDataBase voc_db = new VOCDataBase(context);
        Cursor voc_info = voc_db.getCurrentVOC();

        // set times for last aspect and ingress; add 1 minute to ingress so update will happen just after it occurs
        la.setTimeInMillis(voc_info.getLong(4) * 1000);
        in.setTimeInMillis((voc_info.getLong(1) * 1000) + 60000);

        // round times down to the nearest minute
        now.set(Calendar.MILLISECOND, 0);
        now.set(Calendar.SECOND, 0);
        la.set(Calendar.MILLISECOND, 0);
        la.set(Calendar.SECOND, 0);
        in.set(Calendar.MILLISECOND, 0);
        in.set(Calendar.SECOND, 0);

        // if the last aspect is still in the future
        if ( la.getTimeInMillis() >= now.getTimeInMillis() ) {
            // set an alarm for the exact time of the next closing aspect
            la_svc = PendingIntent.getService(context, 0, la_i, PendingIntent.FLAG_ONE_SHOT );
            la_millis = SystemClock.elapsedRealtime() + la.getTime().getTime() - System.currentTimeMillis();
            m.setExact(AlarmManager.ELAPSED_REALTIME, la_millis, la_svc);
        }

        // set the alarm for the next ingress
        in_svc = PendingIntent.getService(context, 0, in_i, PendingIntent.FLAG_ONE_SHOT );
        in_millis = SystemClock.elapsedRealtime() + in.getTime().getTime() - System.currentTimeMillis();
        m.setExact(AlarmManager.ELAPSED_REALTIME, in_millis, in_svc);

        // There may be multiple widgets active, so update all of them
        for( int appWidgetId : appWidgetIds) {
            updateMoonInfo(context, appWidgetManager, appWidgetId);
        }
    }

    @Override
    public void onDeleted(Context context, int[] appWidgetIds) {}

    @Override
    public void onEnabled(Context context) {
        // Enter relevant functionality for when the first widget is created
    }

    @Override
    public void onDisabled(Context context) {
        // Enter relevant functionality for when the last widget is disabled
    }

    private static native double witm();
    private static native double wits();
}

File: app/src/main/java/com/lapislucera/wheresthemoon/WTMAppWidget.kt
package com.lapislucera.wheresthemoon

import android.app.AlarmManager
import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.widget.RemoteViews
import kotlin.math.max
import kotlin.math.min

/**
 * Home-screen widget showing the current moon phase, sign and degree, and
 * the void-of-course picture: the last aspect the Moon makes before
 * leaving its sign and the time it enters the next one.
 *
 * The widget refreshes on the half-hourly APPWIDGET_UPDATE cycle and also
 * sets exact alarms for the moment of the last aspect (void begins) and
 * one minute after the ingress (void ends), so the "void" line flips at
 * the right wall-clock minute. The 2015 version started a Service from
 * those alarms; modern Android forbids background service starts, so the
 * alarms now send a broadcast back to this receiver instead.
 */
class WTMAppWidget : AppWidgetProvider() {

    override fun onUpdate(context: Context, appWidgetManager: AppWidgetManager, appWidgetIds: IntArray) {
        Log.i(TAG, "onUpdate for widgets ${appWidgetIds.contentToString()}")
        scheduleAlarms(context)
        for (appWidgetId in appWidgetIds) {
            updateMoonInfo(context, appWidgetManager, appWidgetId)
        }
    }

    override fun onAppWidgetOptionsChanged(
        context: Context,
        appWidgetManager: AppWidgetManager,
        appWidgetId: Int,
        newOptions: Bundle,
    ) {
        // Re-render on resize so line text is re-fitted to the new width.
        updateMoonInfo(context, appWidgetManager, appWidgetId)
    }

    override fun onReceive(context: Context, intent: Intent) {
        Log.i(TAG, "onReceive ${intent.action}")
        if (intent.action == ACTION_REFRESH) {
            // Alarm fired at a void boundary: refresh every widget of both
            // kinds (home widget and lock screen complication).
            val manager = AppWidgetManager.getInstance(context)
            val ids = manager.getAppWidgetIds(ComponentName(context, WTMAppWidget::class.java))
            onUpdate(context, manager, ids)
            WTMComplication.updateAll(context)
        } else {
            super.onReceive(context, intent)
        }
    }

    companion object {
        private const val TAG = "WTMAppWidget"
        const val ACTION_REFRESH = "com.lapislucera.wheresthemoon.REFRESH"

        // Sized for readability on modern high-density screens (2026);
        // the original 14/12 was designed for 2015-era phones.
        private const val LABEL_FONT_DP = 22f
        private const val TEXT_FONT_DP = 20f

        // Horizontal space (dp) that is not available to the side columns:
        // widget padding (8+8), center column moon icon (90) + padding
        // (10+10), and a little slack for rounding.
        private const val NON_COLUMN_DP = 130f
        private const val FALLBACK_WIDGET_WIDTH_DP = 320f

        fun updateMoonInfo(context: Context, appWidgetManager: AppWidgetManager, appWidgetId: Int) {
            val views = RemoteViews(context.packageName, R.layout.wtmapp_widget)
            val fonts = WidgetRender.Fonts(context)

            // Current moon position and phase, computed live.
            val jd = Ephemeris.julianDay(System.currentTimeMillis())
            val moonLon = Ephemeris.moonLongitude(jd)
            val sunLon = Ephemeris.sunLongitude(jd)

            val moonIconId = context.resources.getIdentifier(
                MoonDisplay.phaseDrawableName(moonLon, sunLon),
                "drawable",
                context.packageName,
            )
            views.setImageViewResource(R.id.moon_icon, moonIconId)

            views.setImageViewBitmap(
                R.id.moon_position,
                WidgetRender.lineBitmap(
                    context,
                    listOf(
                        WidgetRender.Segment(fonts.kairon, MoonDisplay.SIGNS[MoonDisplay.signIndex(moonLon)] + " "),
                        WidgetRender.Segment(fonts.alegreya, MoonDisplay.dms(moonLon % 30.0)),
                    ),
                    TEXT_FONT_DP,
                ),
            )

            views.setImageViewBitmap(
                R.id.last_aspect_label,
                WidgetRender.lineBitmap(context, listOf(WidgetRender.Segment(fonts.alegreyaBold, "Last Aspect")), LABEL_FONT_DP),
            )
            views.setImageViewBitmap(
                R.id.ingress_label,
                WidgetRender.lineBitmap(context, listOf(WidgetRender.Segment(fonts.alegreyaBold, "Ingress")), LABEL_FONT_DP),
            )

            val voc = VocDatabase(context).getCurrentVoc()
            if (voc == null) {
                // The bundled data has run out; see tools/vocgen in the repo.
                views.setImageViewBitmap(
                    R.id.void_or_not,
                    WidgetRender.lineBitmap(
                        context,
                        listOf(WidgetRender.Segment(fonts.alegreya, context.getString(R.string.voc_data_expired))),
                        TEXT_FONT_DP,
                    ),
                )
                views.setImageViewBitmap(R.id.last_aspect_line, null)
                views.setImageViewBitmap(R.id.ingress_line, null)
                appWidgetManager.updateAppWidget(appWidgetId, views)
                return
            }

            // Times rounded down to the minute, like the original app.
            val ingressMillis = voc.ingress / 60 * 60000
            val asptimeMillis = voc.asptime / 60 * 60000

            // Each info line is one bitmap (glyphs + time), and both lines
            // share one font size, chosen so the wider line fits its
            // column. Rendering them as separate per-part images let the
            // host squeeze one line but not the other, which made the
            // Ingress text look smaller than Last Aspect.
            val aspectLine = listOf(
                WidgetRender.Segment(fonts.kairon, MoonDisplay.ASPECTS[voc.aspect] + " " + MoonDisplay.PLANETS[voc.planet] + " "),
                WidgetRender.Segment(fonts.alegreya, WidgetRender.formatTime(asptimeMillis)),
            )
            val ingressLine = listOf(
                WidgetRender.Segment(fonts.kairon, MoonDisplay.SIGNS[voc.sign] + " "),
                WidgetRender.Segment(fonts.alegreya, WidgetRender.formatTime(ingressMillis)),
            )

            val widgetWidthDp = appWidgetManager.getAppWidgetOptions(appWidgetId)
                .getInt(AppWidgetManager.OPTION_APPWIDGET_MIN_WIDTH)
                .takeIf { it > 0 }?.toFloat() ?: FALLBACK_WIDGET_WIDTH_DP
            val columnPx = WidgetRender.dpToPx(context, max(60f, (widgetWidthDp - NON_COLUMN_DP) / 2f))
            val basePx = WidgetRender.dpToPx(context, TEXT_FONT_DP)
            val widestPx = max(WidgetRender.lineWidth(aspectLine, basePx), WidgetRender.lineWidth(ingressLine, basePx))
            val fittedFontDp = TEXT_FONT_DP * min(1f, columnPx / widestPx)

            views.setImageViewBitmap(R.id.last_aspect_line, WidgetRender.lineBitmap(context, aspectLine, fittedFontDp))
            views.setImageViewBitmap(R.id.ingress_line, WidgetRender.lineBitmap(context, ingressLine, fittedFontDp))

            val now = System.currentTimeMillis()
            val voidText = if (now in asptimeMillis until ingressMillis) {
                context.getString(R.string.moon_is_void)
            } else {
                context.getString(R.string.moon_not_void)
            }
            views.setImageViewBitmap(
                R.id.void_or_not,
                WidgetRender.lineBitmap(context, listOf(WidgetRender.Segment(fonts.alegreya, voidText)), TEXT_FONT_DP),
            )

            appWidgetManager.updateAppWidget(appWidgetId, views)
        }

        /**
         * Set exact alarms to refresh the widgets when the void period
         * begins (last aspect) and just after it ends (ingress + 1 min).
         */
        fun scheduleAlarms(context: Context) {
            val voc = VocDatabase(context).getCurrentVoc() ?: return
            val alarmManager = context.getSystemService(AlarmManager::class.java)
            val now = System.currentTimeMillis()

            val asptimeMillis = voc.asptime / 60 * 60000
            if (asptimeMillis >= now) {
                setAlarm(context, alarmManager, asptimeMillis, requestCode = 0)
            }
            setAlarm(context, alarmManager, (voc.ingress / 60 + 1) * 60000, requestCode = 1)
        }

        private fun setAlarm(context: Context, alarmManager: AlarmManager, triggerAtMillis: Long, requestCode: Int) {
            val pending = PendingIntent.getBroadcast(
                context,
                requestCode,
                Intent(context, WTMAppWidget::class.java).setAction(ACTION_REFRESH),
                PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
            )
            // Exact alarms need a grant on Android 12+; USE_EXACT_ALARM in
            // the manifest provides it automatically on 13+. Fall back to a
            // 10-minute window if it is somehow missing.
            val canExact = Build.VERSION.SDK_INT < Build.VERSION_CODES.S ||
                alarmManager.canScheduleExactAlarms()
            if (canExact) {
                alarmManager.setExactAndAllowWhileIdle(AlarmManager.RTC_WAKEUP, triggerAtMillis, pending)
            } else {
                alarmManager.setWindow(AlarmManager.RTC_WAKEUP, triggerAtMillis, 10 * 60000, pending)
            }
        }
    }
}
